File: algoDat/public/code/vorlesung-07/vorlagen/cpp/LinkedList.cpp
// Copyright 2017, University of Freiburg,
// Chair of Algorithms and Data Structures.
// Author: Hannah Bast <bast@cs.uni-freiburg.de>,
//         Axel Lehmann <lehmann@cs.uni-freiburg.de>.

#include <iostream>
#include <string>
#include <sstream>
#include "./LinkedList.h"
#include "./LinkedListItem.h"

// _____________________________________________________________________________
LinkedList::LinkedList() {
  firstItem = NULL;
  lastItem = NULL;
}

// _____________________________________________________________________________
void LinkedList::insertBefore(LinkedListItem* newItem,
                              LinkedListItem* nextItem) {
  LinkedListItem* previousItem;
  if (nextItem != NULL) {
    previousItem = nextItem->previousItem;
  } else {
    previousItem = lastItem;
  }

  if (nextItem != NULL) {
    nextItem->previousItem = newItem;
  } else {
    lastItem = newItem;
  }

  newItem->previousItem = previousItem;

  if (previousItem != NULL) {
    previousItem->nextItem = newItem;
  } else {
    firstItem = newItem;
  }

  newItem->nextItem = nextItem;
}

// _____________________________________________________________________________
std::string LinkedList::toString() {
  std::ostringstream os;
  os << "first = " << LinkedListItem::idString(firstItem) << ", "
     << "last = " << LinkedListItem::idString(lastItem) << std::endl;
  LinkedListItem* item = firstItem;
  while (item != NULL) {
    os << item->toString();
    if (item->nextItem != NULL) { os << std::endl; }
    item = item->nextItem;
  }
  return os.str();
}

File: algoDat/public/code/vorlesung-07/vorlagen/java/LinkedList.java
// Copyright 2017, University of Freiburg,
// Chair of Algorithms and Data Structures.
// Author: Hannah Bast <bast@cs.uni-freiburg.de>,
//         Axel Lehmann <lehmann@cs.uni-freiburg.de>.

/**
 * Class for a doubly-linked list.
 */
public class LinkedList {

  /**
   * Create an empty list.
   */
  public LinkedList() {
    firstItem = null;
    lastItem = null;
  }

  /**
   * Insert an element into the list, *before* the given item.
   */
  public void insertBefore(LinkedListItem newItem, LinkedListItem nextItem) {
    LinkedListItem previousItem;
    if (nextItem != null) {
      previousItem = nextItem.previousItem;
    } else {
      previousItem = lastItem;
    }

    if (nextItem != null) {
      nextItem.previousItem = newItem;
    } else {
      lastItem = newItem;
    }

    newItem.previousItem = previousItem;

    if (previousItem != null) {
      previousItem.nextItem = newItem;
    } else {
      firstItem = newItem;
    }

    newItem.nextItem = nextItem;
  }

  /**
   * The first item in the list.
   */
  protected LinkedListItem firstItem;

  /**
   * The last item in the list.
   */
  protected LinkedListItem lastItem;

  /**
   * This item as human-readable string.
   */
  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("first = " + LinkedListItem.id(firstItem) + ", ");
    sb.append("last = " + LinkedListItem.id(lastItem) + "\n");
    LinkedListItem item = firstItem;
    while (item != null) {
      sb.append(item.toString());
      if (item.nextItem != null) {
        sb.append("\n");
      }
      item = item.nextItem;
    }
    return sb.toString();
  }
}
